
Overview {#overview_fusion data-orientation=rows data-navmenu="Fusion"}
==================================================================

Row
-------------------------------------

###

```{r fusion_overview}
aux_vars <- fusion %>% filter(NO_FLAGS)
aux_N <- nrow(aux_vars)
aux_label <- "Fusion"
aux_caption <- paste0(aux_N, " fusions TierI-III")

valueBox(aux_label, color = "black", caption = aux_caption, href  = "#main")

rm(aux_vars, aux_N, aux_label, aux_caption)
```

###

```{r fusion_overview_tierI}
aux_vars <- fusion %>% filter(NO_FLAGS & TIER_I)
aux_N <- nrow(aux_vars)
aux_label <- ifelse(aux_N == 1, aux_vars$LABEL, paste0(aux_N, " Tier I"))
aux_caption <- ifelse(aux_N == 0, 
  ifelse(doid %in% ".", "strong clinical significance (No Tier I fusions due to unknown tumor type)", "strong clinical significance"), 
  ifelse(aux_N == 1, "Tier I (strong clinical significance)", 
    paste0(paste(unique(sort(aux_vars$LABEL)), collapse = ", "), " (strong clinical significance)")))

valueBox(aux_label, caption = aux_caption, color = ifelse(aux_N, all_colors["Tier IA"], "grey"), href  = "#plot_fusion")

rm(aux_vars, aux_N, aux_label, aux_caption)
```

###

```{r fusion_overview_tierII}
aux_vars <- fusion %>% filter(NO_FLAGS & TIER_II)
aux_N <- nrow(aux_vars)
aux_label <- ifelse(aux_N == 1, aux_vars$LABEL, paste0(aux_N, " Tier II"))
aux_caption <- ifelse(aux_N == 0, "potential clinical significance", ifelse(
  aux_N == 1, "Tier II (potential clinical significance)", 
  paste0(paste(unique(sort(aux_vars$LABEL)), collapse = ", "), " (potential clinical significance)")))

valueBox(aux_label, caption = aux_caption, color = ifelse(aux_N, all_colors["Tier IIC"], "grey"), href  = "#plot_fusion")

rm(aux_vars, aux_N, aux_label, aux_caption)
```

###

```{r fusion_overview_tierIII}
aux_vars <- fusion %>% filter(NO_FLAGS & TIER_III)
aux_N <- nrow(aux_vars)
aux_label <- paste0(aux_N, " Tier III")
aux_caption <- ifelse(aux_N == 0, "unknown clinical significance", 
  paste0(paste(unique(sort(aux_vars$LABEL)), collapse = ", "), " (unknown clinical significance)"))

valueBox(aux_label, caption = aux_caption, color = ifelse(aux_N, all_colors["Tier III"], "grey"), href  = "#plot_fusion")

rm(aux_vars, aux_N, aux_label, aux_caption)
```



Row {data-height=350}
-------------------------------------

### Fusion CALLING

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r all_colors["OK"]`">&nbsp;PASS&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r nrow(fusion_called)`</font></b><br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r all_colors["FAIL"]`">&nbsp;LowSupport (AD<`r fusion_calling_min_ad` or FFPM<`r fusion_calling_min_ffpm`)&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r nrow(fusion %>% filter(LowSupport))`</font></b><br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r all_colors["FAIL"]`">&nbsp;Normal&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r nrow(fusion %>% filter(Normal))`</font></b><br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r all_colors["FAIL"]`">&nbsp;NoInSilicoValid&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r nrow(fusion %>% filter(!InSilicoValid))`</font></b><br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r all_colors["FAIL"]`">&nbsp;RTartifact&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r nrow(fusion %>% filter(RTartifact))`</font></b><br>



### Fusion FLAGS

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r all_colors["OK"]`">&nbsp;OK (no flags)&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r nrow(fusion %>% filter(NO_FLAGS))`</font></b><br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r all_colors["FAIL"]`">&nbsp;LowAD (<`r fusion_flag_min_ad`)&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r nrow(fusion_called %>% filter(LowAD %in% "TRUE"))`</font></b><br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r all_colors["FAIL"]`">&nbsp;LowDP (<`r fusion_flag_min_dp`)&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r nrow(fusion_called %>% filter(LowDP %in% "TRUE"))`</font></b><br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r all_colors["FAIL"]`">&nbsp;LowNonFused (<`r fusion_flag_min_ad_nonfused`)&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r nrow(fusion_called %>% filter(LowNonFused %in% "TRUE"))`</font></b><br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r all_colors["FAIL"]`">&nbsp;LowVAF (<`r fusion_flag_min_vaf*100`%)&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r nrow(fusion_called %>% filter(LowVAF %in% "TRUE"))`</font></b><br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r all_colors["FAIL"]`">&nbsp;Unknown&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r nrow(fusion_called %>% filter(UNKNOWN))`</font></b><br>



### Clinical status (AMP/ASCO/CAP)

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r all_colors["Tier IA"]`">&nbsp;Tier I&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r nrow(fusion_called %>% filter(TIER_I))`</font></b><br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r all_colors["Tier IIC"]`">&nbsp;Tier II&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r nrow(fusion_called %>% filter(TIER_II))`</font></b><br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r all_colors["Tier III"]`">&nbsp;Tier III&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r nrow(fusion_called %>% filter(TIER_III))`</font></b><br>

* <font style="color:white; font-size: 125%; font-weight: bold; background-color:`r all_colors["Tier IV"]`">&nbsp;Tier IV&nbsp;&nbsp;</font></a> <b><font style="font-size: 125%">:&nbsp;`r nrow(fusion_called %>% filter(TIER_IV))`</font></b><br>



### Tier I/II fusions

```{r fusion_overview_tierI-II_ok}
aux_vars <- fusion_called %>% filter(CLIN_RELEVANT & NO_FLAGS)
aux_N <- nrow(aux_vars)
aux_label <- ifelse(aux_N, paste(unique(sort(aux_vars$LABEL)), collapse = " <br>"), "ND")
aux_color <- ifelse(aux_N, "white", "black")
aux_bg <- ifelse(aux_N, all_colors["Tier IA"], "white")
```

#### **OK** (N=`r aux_N`)

* <font style="color:`r aux_color`; font-size: 125%; background-color:`r aux_bg`">&nbsp;`r aux_label`&nbsp;</font></b><br>

```{r}
rm(aux_vars, aux_N, aux_label, aux_color, aux_bg)
```


```{r fusion_overview_tierI-II_flagged}
aux_vars <- fusion_called %>% filter(CLIN_RELEVANT & !NO_FLAGS)
aux_N <- nrow(aux_vars)
aux_label <- ifelse(aux_N, paste(unique(sort(aux_vars$LABEL)), collapse = " <br>"), "ND")
aux_color <- ifelse(aux_N, "white", "black")
aux_bg <- ifelse(aux_N, all_colors["Tier IA"], "white")
```

#### **Flagged** (N=`r aux_N`)

* <font style="color:`r aux_color`; font-size: 125%; background-color:`r aux_bg`">&nbsp;`r aux_label`&nbsp;</font></b><br>

```{r}
rm(aux_vars, aux_N, aux_label, aux_color, aux_bg)
```



### Fusions

```{r fusion_overview_ok}
aux_vars <- fusion_called %>% filter(NO_FLAGS)
aux_N <- nrow(aux_vars)
aux_label <- ifelse(aux_N, paste(unique(sort(aux_vars$LABEL)), collapse = " <br>"), "ND")
aux_color <- ifelse(aux_N, "white", "black")
aux_bg <- ifelse(aux_N, all_colors["OK"], "white")
```

#### **OK** (N=`r aux_N`)

* <font style="color:`r aux_color`; font-size: 125%; background-color:`r aux_bg`">&nbsp;`r aux_label`&nbsp;</font></b><br>

```{r}
rm(aux_vars, aux_N, aux_label, aux_color, aux_bg)
```


```{r fusion_overview_flagged}
aux_vars <- fusion_called %>% filter(!NO_FLAGS)
aux_N <- nrow(aux_vars)
aux_label <- ifelse(aux_N, paste(unique(sort(aux_vars$LABEL)), collapse = " <br>"), "ND")
aux_color <- ifelse(aux_N, "white", "black")
aux_bg <- ifelse(aux_N, all_colors["FAIL"], "white")
```

#### **Flagged** (N=`r aux_N`)

* <font style="color:`r aux_color`; font-size: 125%; background-color:`r aux_bg`">&nbsp;`r aux_label`&nbsp;</font></b><br>

```{r}
rm(aux_vars, aux_N, aux_label, aux_color, aux_bg)
```



Row {.tabset .tabset-fade}
-------------------------------------

::: {.header-text}
Clinical evidences (CIViC resource)
:::

### **Predictive**

```{r fusion_overview_predictive}
aux_df <- civic_predictive %>% filter(MOLECULAR_ID %in% fusion_called$CIViC_MOLECULAR_ID)
aux_kable <- kable(aux_df, align = "c") %>% 
  kable_styling(bootstrap_options = c("striped", "condensed"), fixed_thead = TRUE, font_size = 16)

if(nrow(aux_df)){column_spec(aux_kable, 1, bold = TRUE)} else {aux_kable}

rm(aux_df, aux_kable)
```

### **Prognostic**

```{r fusion_overview_prognostic}
aux_df <- civic_prognostic %>% filter(MOLECULAR_ID %in% fusion_called$CIViC_MOLECULAR_ID)
aux_kable <- kable(aux_df, align = "c") %>% 
  kable_styling(bootstrap_options = c("striped", "condensed"), fixed_thead = TRUE, font_size = 16)

if(nrow(aux_df)){column_spec(aux_kable, 1, bold = TRUE)} else {aux_kable}

rm(aux_df, aux_kable)
```

### **Diagnostic**

```{r fusion_overview_diagnostic}
aux_df <- civic_diagnostic %>% filter(MOLECULAR_ID %in% fusion_called$CIViC_MOLECULAR_ID)
aux_kable <- kable(aux_df, align = "c") %>% 
  kable_styling(bootstrap_options = c("striped", "condensed"), fixed_thead = TRUE, font_size = 16)

if(nrow(aux_df)){column_spec(aux_kable, 1, bold = TRUE)} else {aux_kable}

rm(aux_df, aux_kable)
```



Gene visualization {#plot_fusion data-orientation=rows data-navmenu="Fusion"}
==================================================================

Row {.tabset .tabset-fade}
-------------------------------------

```{r fusion_plot, results = "asis"}
for(n in seq_len(nrow(fusion_plot))) {
  id <- paste0(
    gsub("::", "-", fusion_plot[n, "FUSION_SHORT"]), "_", 
    gsub(":", "_", fusion_plot[n, "BREAKPOINT_A"]), "-", gsub(":", "_", fusion_plot[n, "BREAKPOINT_B"]))
  fusion_name <- fusion_plot[n, "FUSION_NAME"]
  aux_plot <- list.files(pattern = paste0(sample, "_RNA_Fusion_", id, ".png"))
  
  if(length(aux_plot)){cat(sprintf(paste0("\n### **", fusion_name, "**\n![](%s)\n"), aux_plot))}
  
  rm(id, fusion_name, aux_plot)
}
rm(n)
```

```{r fusion_noplot, results = 'asis', eval = nrow(fusion_plot) == 0}
cat("\n* **NO** Fusion plots.\n")
```



Datatable {data-orientation=columns data-navmenu="Fusion"}
==================================================================

Column {.sidebar data-width=250}
-----------------------------------------------

```{r fusion_filters, eval = nrow(fusion_table) != 0}
shared_data <- SharedData$new(fusion_table)

filter_checkbox("FUSION_OK", "OK (no flags)", shared_data, ~NO_FLAGS, columns = 2, allLevels = TRUE)
filter_checkbox("FUSION_CLIN_STATUS", "CLINICAL status", shared_data, ~CLIN_STATUS, columns = 2, allLevels = TRUE)
filter_checkbox("FUSION_WHITELIST", "WHITELIST fusion", shared_data, ~WHITELIST_FUSION, columns = 2, allLevels = TRUE)
filter_checkbox("FUSION_WHITELIST_GENE", "WHITELIST gene", shared_data, ~WHITELIST_GENE, columns = 2, allLevels = TRUE)
filter_checkbox("FUSION_MitelmanDB", "MitelmanDB", shared_data, ~MitelmanDB, columns = 2, allLevels = TRUE)
filter_checkbox("FUSION_GENIE", "GENIE", shared_data, ~GENIE, columns = 2, allLevels = TRUE)
filter_checkbox("FUSION_TYPE", "Model fusion type", shared_data, ~TYPE, columns = 2, allLevels = TRUE)
filter_checkbox("FUSION_LowAD", "LowAF", shared_data, ~LowAD, columns = 2, allLevels = TRUE)
filter_checkbox("FUSION_LowDP", "LowAF", shared_data, ~LowDP, columns = 2, allLevels = TRUE)
filter_checkbox("FUSION_LowNonFused", "LowAF", shared_data, ~LowNonFused, columns = 2, allLevels = TRUE)
filter_checkbox("FUSION_LowVAF", "LowAF", shared_data, ~LowVAF, columns = 2, allLevels = TRUE)
filter_checkbox("FUSION_Unknown", "No fusion DBs", shared_data, ~UNKNOWN, columns = 2, allLevels = TRUE)

filter_select("FUSION_GENE_A", "GENE A", shared_data, ~GENE_A, multiple = TRUE)
filter_select("FUSION_GENE_B", "GENE B", shared_data, ~GENE_B, multiple = TRUE)
filter_select("FUSION_FLAGS", "FLAGS", shared_data, ~FLAGS, multiple = TRUE)

filter_slider("FUSION_AD", "Fusion-supporting reads", shared_data, ~AD, ticks = FALSE)
filter_slider("FUSION_AF", "Allelic fraction (AF)", shared_data, ~AF, min = 0, max = 1, step = 0.1, ticks = TRUE)
filter_slider("FUSION_DP", "Sequencing depth (DP)", shared_data, ~DP, ticks = FALSE)
filter_slider("FUSION_AF_SAMPLES", "Sample frequency", shared_data, ~AF_SAMPLES, min = 0, max = 1, step = 0.1, ticks = TRUE)

```


Column
---------------------------------------------

### Fusions

```{r fusion_table, eval = nrow(fusion_table) != 0}
datatable(
  shared_data, class = "display", extensions = c("Buttons", "ColReorder"), rownames = FALSE,
  options = list(
    bPaginate = FALSE, searchHighlight = TRUE, dom = "BfrtiR", buttons = c("copy", "csv","excel"),
    initComplete = JS(
    "function(settings, json) {",
    paste0("$(this.api().table().header()).css({'background-color': '", all_colors["primary"], "', 'color': '#fff'});"),
    "}"))) %>%
  formatStyle("CLIN_RELEVANT", target = "row", fontWeight = styleEqual("TRUE", "bold"), color = styleEqual("TRUE", all_colors["CLIN_RELEVANT"])) %>%
  formatStyle("FLAGS", "NO_FLAGS", fontWeight = styleEqual(c("TRUE", "FALSE"), rep("bold", 2)), 
    color = styleEqual(c("TRUE", "FALSE"), c(all_colors["OK"], all_colors["FAIL"]))) %>%
  formatStyle("CLIN_STATUS", fontWeight = styleEqual(clinical_levels, rep("bold", length(clinical_levels))),
    color = styleEqual(clinical_levels, all_colors[clinical_levels])) %>%
  formatStyle("CALLING", "PASS", fontWeight = styleEqual(c("TRUE", "FALSE"), rep("bold", 2)), 
    color = styleEqual(c("TRUE", "FALSE"), c(all_colors["OK"], all_colors["FAIL"]))) %>%
  formatStyle("CLIN_RELEVANT", fontWeight = styleEqual("TRUE", "bold")) %>%
  formatStyle("WHITELIST_FUSION", target = "row", fontWeight = styleEqual("TRUE", "bold")) %>%
  formatStyle("WHITELIST_GENE", fontWeight = styleEqual("TRUE", "bold")) %>%
  formatStyle("MitelmanDB", color = styleEqual("TRUE", all_colors["OK"])) %>%
  formatStyle("GENIE", color = styleEqual("TRUE", all_colors["OK"])) %>%
  formatStyle("PASS", color = styleEqual(c("TRUE", "FALSE"), c(all_colors["OK"], all_colors["FAIL"]))) %>%
  formatStyle("LowSupport", color = styleEqual(c("TRUE", "FALSE"), c(all_colors["FAIL"], all_colors["OK"]))) %>%
  formatStyle("Normal", color = styleEqual(c("TRUE", "FALSE"), c(all_colors["FAIL"], all_colors["OK"]))) %>%
  formatStyle("RTartifact", color = styleEqual(c("TRUE", "FALSE"), c(all_colors["FAIL"], all_colors["OK"]))) %>%
  formatStyle("InSilicoValid", color = styleEqual(c("TRUE", "FALSE"), c(all_colors["OK"], all_colors["FAIL"]))) %>%
  formatStyle("NO_FLAGS", color = styleEqual(c("TRUE", "FALSE"), c(all_colors["OK"], all_colors["FAIL"]))) %>%
  formatStyle("LowAD", color = styleEqual(c("TRUE", "FALSE"), c(all_colors["FAIL"], all_colors["OK"]))) %>%
  formatStyle("LowDP", color = styleEqual(c("TRUE", "FALSE"), c(all_colors["FAIL"], all_colors["OK"]))) %>%
  formatStyle("LowNonFused", color = styleEqual(c("TRUE", "FALSE"), c(all_colors["FAIL"], all_colors["OK"]))) %>%
  formatStyle("LowVAF", color = styleEqual(c("TRUE", "FALSE"), c(all_colors["FAIL"], all_colors["OK"]))) %>%
  formatStyle("UNKNOWN", color = styleEqual(c("TRUE", "FALSE"), c(all_colors["FAIL"], all_colors["OK"]))) %>%
  formatStyle("LargeAnchorSupport", fontWeight = styleEqual("TRUE", "bold")) %>%
  formatStyle("RefSpliceSite", fontWeight = styleEqual("TRUE", "bold")) %>%
  formatStyle("ONCOGENE_A", fontWeight = styleEqual("TRUE", "bold")) %>%
  formatStyle("TSG_A", fontWeight = styleEqual("TRUE", "bold")) %>%
  formatStyle("DRIVER_A", fontWeight = styleEqual("TRUE", "bold")) %>%
  formatStyle("CANONICAL_DRIVER_A", fontWeight = styleEqual("TRUE", "bold")) %>%
  formatStyle("ONCOGENE_B", fontWeight = styleEqual("TRUE", "bold")) %>%
  formatStyle("TSG_B", fontWeight = styleEqual("TRUE", "bold")) %>%
  formatStyle("DRIVER_B", fontWeight = styleEqual("TRUE", "bold")) %>%
  formatStyle("CANONICAL_DRIVER_B", fontWeight = styleEqual("TRUE", "bold"))

rm(shared_data)
```


```{r, results = 'asis', eval = nrow(fusion_table) == 0}
cat("\n* **NO** Fusion results.\n")
```

